### PR TITLE
Whole-core GCMR Neutronic Model

### PR DESCRIPTION
**Pull request by aabdelhameedd**
This PR is related to issue #374

### CODE REVIEW

**Comment by nstauff** [comment]
Thanks @aabdelhameedd  - I added Shikhar as reviewer. After our review, we should push this for review to VTB developers.

**Review by nstauff — APPROVED** [review]

**Comment by shikhar413** [comment]
@aabdelhameedd I’m on vacation this week, let me get back to you with a review next week

File: doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md
Comment (shikhar413) [inline doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md:5]: GC-MR here, but GCMR in the title

File: doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md
Comment (shikhar413) [inline doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md:17]: comprises of

Comment (shikhar413) [inline doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md:17]: Maybe more explicitly define what the differences between Assemblies A, B, and C, since it may not be clear from just looking at the images

File: doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md
Comment (shikhar413) [inline doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md:54]: Reference reactor module

@article{shemon2023reactor,
 author = {Emily Shemon and Yinbin Miao and Shikhar Kumar and Kun Mo and Yeon Sang Jung and Aaron Oaks and Scott Richards and Guillaume Giudicelli and Logan Harbour and Roy Stogner},
  title = {MOOSE Reactor Module: An Open-Source Capability for Meshing Nuclear Reactor Geometries},
journal = {Nuclear Science and Engineering},
 volume = {0},
 number = {0},
  pages = {1-25},
  year  = {2023},
    doi = {10.1080/00295639.2022.2149231},
    URL = {https://doi.org/10.1080/00295639.2022.2149231},
 eprint = {https://doi.org/10.1080/00295639.2022.2149231}
}

Comment (shikhar413) [inline doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md:54]: GCMR*

Comment (shikhar413) [inline doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md:54]: What does "reasonable computational demands" mean? I would leave out this phrase unless you want to provide specifics for comparison

File: doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md
Comment (shikhar413) [inline doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md:69]: The process of defining the Griffin model begins by*

Comment (shikhar413) [inline doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md:69]: which were then converted into *an* XML-format cross-section file

Comment (shikhar413) [inline doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md:69]: I would remove the word meticulous

Comment (shikhar413) [inline doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md:69]: discontinuous finite element mesh (DFEM)

File: doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md
Comment (shikhar413) [inline doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md:80]: [number_of_cores]

File: doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md
Comment (shikhar413) [inline doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md:82]: It should be noted

File: doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md
Comment (shikhar413) [inline doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md:89]: in the SN calculation*

Comment (shikhar413) [inline doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md:89]: DFEM-SN (n_polar=2, n_azimuthal=3) with CMFD*

n case there is confusion about what (2, 3) means

File: microreactors/gcmr/core/MESH/Griffin_mesh.i
Comment (shikhar413) [inline microreactors/gcmr/core/MESH/Griffin_mesh.i:16]: Remove commented lined, same for other mesh blocks

File: microreactors/gcmr/core/MESH/Griffin_mesh.i
Comment (shikhar413) [inline microreactors/gcmr/core/MESH/Griffin_mesh.i:30]: This is a default parameter and can be removed, same for other mesh blocks

File: microreactors/gcmr/core/MESH/Griffin_mesh.i
Comment (shikhar413) [inline microreactors/gcmr/core/MESH/Griffin_mesh.i:87]: tabbing is off

File: microreactors/gcmr/core/MESH/Griffin_mesh.i
Comment (shikhar413) [inline microreactors/gcmr/core/MESH/Griffin_mesh.i:154]: Make this into a hexagonal pattern similar to the top half of the pattern?

File: microreactors/gcmr/core/MESH/Griffin_mesh.i
Comment (shikhar413) [inline microreactors/gcmr/core/MESH/Griffin_mesh.i:157]: Remove if not needed

File: microreactors/gcmr/core/MESH/Griffin_mesh.i
Comment (shikhar413) [inline microreactors/gcmr/core/MESH/Griffin_mesh.i:160]: Check tabbing

File: microreactors/gcmr/core/MESH/Griffin_mesh.i
Comment (shikhar413) [inline microreactors/gcmr/core/MESH/Griffin_mesh.i:186]: Check tabbing

File: doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md
Comment (shikhar413) [inline doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md:77]: This section makes it seem like the simulation can be run out of the box, but the mesh needs to be generated first since it is imported with FileMeshGenerator. I would describe mesh generation first and then the simulation. Alternatively, you can also run the simulation and mesh generation as one command too by removing the [Mesh] block in Griffin_stead_state.i

File: microreactors/gcmr/core/Neutronics/Griffin_steady_state.i
Comment (shikhar413) [inline microreactors/gcmr/core/Neutronics/Griffin_steady_state.i:18]: Can we make this block ID numbering to match the mapping of cross section IDs? It's really difficult to follow the mesh-material mapping with all of these numbers

File: microreactors/gcmr/core/Neutronics/Griffin_steady_state.i
Comment (shikhar413) [inline microreactors/gcmr/core/Neutronics/Griffin_steady_state.i:52]: Remove

File: microreactors/gcmr/core/Neutronics/Griffin_steady_state.i
Comment (shikhar413) [inline microreactors/gcmr/core/Neutronics/Griffin_steady_state.i:66]: Is this block needed?

File: microreactors/gcmr/core/Neutronics/Griffin_steady_state.i
Comment (shikhar413) [inline microreactors/gcmr/core/Neutronics/Griffin_steady_state.i:105]: What does this comment mean?

File: microreactors/gcmr/core/Neutronics/Griffin_steady_state.i
Comment (shikhar413) [inline microreactors/gcmr/core/Neutronics/Griffin_steady_state.i:115]: Is this line needed

File: microreactors/gcmr/core/Neutronics/Griffin_steady_state.i
Comment (shikhar413) [inline microreactors/gcmr/core/Neutronics/Griffin_steady_state.i:152]: Remove ./ and ../

File: microreactors/gcmr/core/Neutronics/Griffin_steady_state.i
Comment (shikhar413) [inline microreactors/gcmr/core/Neutronics/Griffin_steady_state.i:151]: I think it'd be useful to describe what these blocks do in the tutorial documentation

File: microreactors/gcmr/core/Neutronics/tests
Comment (GiudGiud) [inline microreactors/gcmr/core/Neutronics/tests:5]: why _alone? are there other inputs to come?

File: microreactors/gcmr/core/Neutronics/tests
Comment (GiudGiud) [inline microreactors/gcmr/core/Neutronics/tests:10]: I dont see a multiapps block?

File: microreactors/gcmr/core/Neutronics/Griffin_steady_state.i
Comment (GiudGiud) [inline microreactors/gcmr/core/Neutronics/Griffin_steady_state.i:45]: I would re-order the parameters and group them (with a comment) for readability.
maybe 2 groups?
- richardson iterations
- CMFD transport acceletation

File: microreactors/gcmr/core/Neutronics/tests
Comment (GiudGiud) [inline microreactors/gcmr/core/Neutronics/tests:1]: you ll need a first 'RunApp' test that generates the mesh

then get rid of parallel scheduling because that first test must run first

File: doc/content/microreactors/gcmr/index.md
Comment (GiudGiud) [inline doc/content/microreactors/gcmr/index.md:36]:
```suggestion
# Gas-Cooled Whole Core neutronics
```

File: doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md
Comment (GiudGiud) [inline doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md:2]: add tags if you want this model to show up separately in the automated index

File: doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md
Comment (GiudGiud) [inline doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md:5]: I would say the GCMR model, developed at ANL,

File: doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md
Comment (GiudGiud) [inline doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md:35]: Radial or axial? do you want to say enrichment zones?

File: doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md
Comment (GiudGiud) [reply doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md:54]: let s add that to the VTB bib file if it s not in there already:?

File: doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md
Comment (GiudGiud) [reply doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md:82]: put this command first

**Review by GiudGiud — CHANGES_REQUESTED** [review]

File: doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md
Comment (aabdelhameedd) [reply doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md]: Added.

Comment (aabdelhameedd) [reply doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md]: Done.

File: doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md
Comment (aabdelhameedd) [reply doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md:54]: Added to VTB bib

Comment (aabdelhameedd) [reply doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md:54]: Changed.

File: doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md
Comment (aabdelhameedd) [reply doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md]: Done.

File: microreactors/gcmr/core/Neutronics/Griffin_steady_state.i
Comment (aabdelhameedd) [reply microreactors/gcmr/core/Neutronics/Griffin_steady_state.i:46]: Done.

File: doc/content/microreactors/gcmr/index.md
Comment (aabdelhameedd) [reply doc/content/microreactors/gcmr/index.md]: Done.

File: doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md
Comment (aabdelhameedd) [reply doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md]: Clarified. 

We have 3 radial zones ( different FA designs )

File: doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md
Comment (aabdelhameedd) [reply doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md]: It looks the same sentence.

File: doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md
Comment (aabdelhameedd) [reply doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md]: Removed.

File: doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md
Comment (aabdelhameedd) [reply doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md]: Modified.

Comment (aabdelhameedd) [reply doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md]: Removed.

File: doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md
Comment (aabdelhameedd) [reply doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md]: Added.

File: doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md
Comment (aabdelhameedd) [reply doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md]: Modified.

File: microreactors/gcmr/core/MESH/Griffin_mesh.i
Comment (aabdelhameedd) [reply microreactors/gcmr/core/MESH/Griffin_mesh.i]: Removed.

File: microreactors/gcmr/core/MESH/Griffin_mesh.i
Comment (aabdelhameedd) [reply microreactors/gcmr/core/MESH/Griffin_mesh.i]: Modified.

File: microreactors/gcmr/core/MESH/Griffin_mesh.i
Comment (aabdelhameedd) [reply microreactors/gcmr/core/MESH/Griffin_mesh.i]: Removed comment.

File: microreactors/gcmr/core/MESH/Griffin_mesh.i
Comment (aabdelhameedd) [reply microreactors/gcmr/core/MESH/Griffin_mesh.i]: Tabbing is corrected.

File: microreactors/gcmr/core/MESH/Griffin_mesh.i
Comment (aabdelhameedd) [reply microreactors/gcmr/core/MESH/Griffin_mesh.i]: Done.

File: microreactors/gcmr/core/MESH/Griffin_mesh.i
Comment (aabdelhameedd) [reply microreactors/gcmr/core/MESH/Griffin_mesh.i]: Tabbing is corrected.

File: microreactors/gcmr/core/Neutronics/Griffin_steady_state.i
Comment (aabdelhameedd) [reply microreactors/gcmr/core/Neutronics/Griffin_steady_state.i:69]: This block is unnecessary if everything functions as intended, but it can provide additional information for identifying issues in the event they arise.

File: microreactors/gcmr/core/Neutronics/Griffin_steady_state.i
Comment (aabdelhameedd) [reply microreactors/gcmr/core/Neutronics/Griffin_steady_state.i]: Removed.

File: microreactors/gcmr/core/Neutronics/Griffin_steady_state.i
Comment (aabdelhameedd) [reply microreactors/gcmr/core/Neutronics/Griffin_steady_state.i]: Removed.

Comment (aabdelhameedd) [reply microreactors/gcmr/core/Neutronics/Griffin_steady_state.i]: Removed.

File: microreactors/gcmr/core/Neutronics/tests
Comment (aabdelhameedd) [reply microreactors/gcmr/core/Neutronics/tests:5]: Yes, we are planning to add Multiphysics later.

Comment (aabdelhameedd) [reply microreactors/gcmr/core/Neutronics/tests:5]: Yes, we are planning to add Multiphysics later.

File: microreactors/gcmr/core/Neutronics/tests
Comment (aabdelhameedd) [reply microreactors/gcmr/core/Neutronics/tests:10]: We will add Multiphysics later.

File: microreactors/gcmr/core/Neutronics/Griffin_steady_state.i
Comment (aabdelhameedd) [reply microreactors/gcmr/core/Neutronics/Griffin_steady_state.i:18]: Thank you for your suggestion. I think it can be followed by examining the rename_blocks in the mesh input. But we can improve later too.

File: microreactors/gcmr/core/Neutronics/Griffin_steady_state.i
Comment (aabdelhameedd) [reply microreactors/gcmr/core/Neutronics/Griffin_steady_state.i]: > Remove ./ and ../

removed

File: microreactors/gcmr/core/Neutronics/Griffin_steady_state.i
Comment (aabdelhameedd) [reply microreactors/gcmr/core/Neutronics/Griffin_steady_state.i:152]: Thank you for the suggestion. Additional details can be incorporated into the tutorial documentation at a later time, as only one week's worth of funding was allocated for documenting this model.

File: doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md
Comment (aabdelhameedd) [reply doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md]: Indeed, you're correct. I've revised the section to commence with a description of mesh generation first.

File: microreactors/gcmr/core/Neutronics/Griffin_steady_state.i
Comment (aabdelhameedd) [reply microreactors/gcmr/core/Neutronics/Griffin_steady_state.i]: Done.

File: microreactors/gcmr/core/Neutronics/tests
Comment (aabdelhameedd) [reply microreactors/gcmr/core/Neutronics/tests:1]: Added a RunApp to generate the mesh,

File: doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md
Comment (shikhar413) [reply doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md]: There’s an additional blank line there to make sure markdown registers the header and new line properly

File: microreactors/gcmr/core/Neutronics/Griffin_steady_state.i
Comment (shikhar413) [reply microreactors/gcmr/core/Neutronics/Griffin_steady_state.i:151]: I would say this PR should probably be left under review until additional funding is available then. I personally think this sort of work should be in a relative state of readiness for the general user to be able to understand unambiguously, otherwise it falls on the maintainers of the VTB code owners to address any gaps in the documentation/ input files should any issues related to unresolved comments during the review stage arise.

However, I’ll let @GiudGiud make the final call on whether this work is ready to be merged

File: microreactors/gcmr/core/Neutronics/Griffin_steady_state.i
Comment (shikhar413) [reply microreactors/gcmr/core/Neutronics/Griffin_steady_state.i:68]: I’ll defer to @GiudGiud on this since I’m not sure if there’s a repository standard but if these debugging commands are actually something the user would need to concern themselves with they should be described in the tutorial documentation instead and removed here

File: microreactors/gcmr/core/Neutronics/Griffin_steady_state.i
Comment (shikhar413) [reply microreactors/gcmr/core/Neutronics/Griffin_steady_state.i:18]: I would push for using the names in Mesh/fmg_id/subdomains here then. @GiudGiud is there any standard that VTB enforces to make mesh block / material ID mapping clear to the user?

File: doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md
Comment (aabdelhameedd) [reply doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md]: Added.

File: doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md
Comment (aabdelhameedd) [reply doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md]: I see. Yes, an additional blank line was added.

File: microreactors/gcmr/core/MESH/Griffin_mesh.i
Comment (aabdelhameedd) [reply microreactors/gcmr/core/MESH/Griffin_mesh.i]: I've removed the default parameters.

File: doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md
Comment (GiudGiud) [reply doc/content/microreactors/gcmr/GCMR_Core_Neutronics.md:2]: done

**Review by GiudGiud — APPROVED** [review]
As long as it passes tests this should be good to go

**Comment by loganharbour** [comment]
please stop invalidating until it's moved to the new client. this has killed 6 boxes and every time you invalidate it kills another.

**Comment by aabdelhameedd** [comment]
> @aabdelhameedd what is the expected memory consumption of this Griffin input? We are going OOM on a pretty beefy machine Logan set up specially for this

I've used 16 processors on the ANL's VTR machine and it ran well.  Please see the framework information below. I've also attached the complete the running summary for your reference. 
[Griffin_NewGCMR_FY24.txt](https://github.com/idaholab/virtual_test_bed/files/15249922/Griffin_NewGCMR_FY24.txt)

Framework Information:
MOOSE Version:           git commit 34d2c8b5aa on 2022-12-02
LibMesh Version:         
PETSc Version:           3.16.6
SLEPc Version:           3.16.2
Current Time:            Tue Jan 23 20:37:58 2024
Executable Timestamp:    Sat Dec  3 00:52:11 2022

Parallelism:
  Num Processors:          16
  Num Threads:             1

Mesh: 
  Parallel Type:           replicated
  Mesh Dimension:          3
  Spatial Dimension:       3
  Nodes:                   
    Total:                 1093375
    Local:                 70479
    Min/Max/Avg:           65951/70479/68335
  Elems:                   
    Total:                 1228608
    Local:                 76675
    Min/Max/Avg:           75960/77698/76788
  Num Subdomains:          44
  Num Partitions:          16
  Partitioner:             metis

Nonlinear System:
  Num DOFs:                2594820096
  Num Local DOFs:          161937600
  Variables:               { "aflux_g0_0" "aflux_g0_1" "aflux_g0_2" "aflux_g0_3" "aflux_g0_4" ... "aflux_g10_43" 
                             "aflux_g10_44" "aflux_g10_45" "aflux_g10_46" "aflux_g10_47" } 
  Finite Element Types:    "MONOMIAL" 
  Approximation Orders:    "FIRST" 

Auxiliary System:
  Num DOFs:                497586240
  Num Local DOFs:          31053375

**Comment by GiudGiud** [comment]
umm. I dont know if we are going to be able to find a machine with this much memory to test this.
We may relegate this to HPC testing if we cant. 
@loganharbour for advice / awareness

**Comment by GiudGiud** [comment]
Testing machines are down but inputs are no longer tested anyway. So does not matter

**Comment by GiudGiud** [comment]
thanks for the contribution @aabdelhameedd

**Comment by aabdelhameedd** [comment]
> thanks for the contribution @aabdelhameedd

Thank you @GiudGiud for your efforts!

**Pull request merged**
